src/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.AddSub.cs

```diff
@@ -5,6 +5,53 @@ namespace System.Numerics
 {
     internal static partial class BigIntegerCalculator
     {
+        public static uint[] Add(uint[] left, uint right)
+        {
+            Debug.Assert(left != null);
+            Debug.Assert(left.Length >= 1);
+
+            // Executes the addition for one big and one 32-bit integer.
+            // Thus, we've just rather simplified code than below.
+
+            uint[] bits = new uint[left.Length + 1];
+
+            long digit = (long)left[0] + right;
+            bits[0] = (uint)digit;
+            long carry = digit >> 32;
+
+            for (int i = 1; i < left.Length; i++)
+            {
+                digit = left[i] + carry;
+                bits[i] = (uint)digit;
+                carry = digit >> 32;
+            }
+            bits[left.Length] = (uint)carry;
+
+            return bits;
+        }
+
+        [SecuritySafeCritical]
+        public unsafe static uint[] Add(uint[] left, uint[] right)
+        {
+            Debug.Assert(left != null);
+            Debug.Assert(right != null);
+            Debug.Assert(left.Length >= right.Length);
+            
+            // Switching to unsafe pointers helps sparing
+            // some nasty index calculations...
```

Comment:
Using pointer arithmetic? It's improving the performance of the more complex operations (Multiply, Divide) a bit, but only around 10%, if I remember correctly (haven't the numbers anymore). Which adds up for the even more complex ones, like `ModPow`.

For Add/Subtract it's just convenient, because we can use the code within other algorithms, which need to perform these calculations only on parts of a number. Multiply splits a number, which is very straightforward using a pointer. Divide compares "shifted" parts, which is quite easy using a pointer too. Thus, the main benefit is we don't have to write (many) different versions of this, depending on shifts / truncates / whatever.